Search filters and sorting

andres-holguin opened the pull request:
This PR brings the changes from PR #58 and #62 to the master branch. They were originally made to the components branch. Anyone who has reviewed #58 and #62 has reviewed these changes already.

Aradhya-B committed:
Add user service class

Aradhya-B committed:
Overload generic service methods to handle string ids

Aradhya-B committed:
Change long ids to String ids to reflect schema change

Aradhya-B committed:
Merge branch 'auth' into aradhya

Aradhya-B committed:
Merge branch 'search' into aradhya

Aradhya-B committed:
Merge branch 'user-service' into aradhya

Aradhya-B committed:
Merge branch 'blobstore-servlet' into aradhya

Aradhya-B committed:
Merge branch 'components' into aradhya

keelywan committed:
Merge remote-tracking branch 'origin/search' into search-service

Aradhya-B committed:
Add label service class

Aradhya-B committed:
Add misc note label service class

Aradhya-B committed:
Add test page that holds the upload form vue component

Aradhya-B committed:
Add missing '&' in query string for id token and make the method POST

Aradhya-B committed:
Remove newline at end of file

Aradhya-B committed:
Change 'other' type in label to 'misc'

Aradhya-B committed:
Change upload-notes path to handle-notes path

Aradhya-B committed:
Add database interfacing to add notes and labels from the HandleNotesServlet to DB

andres-holguin committed:
Moved files into respective folders

All html files with a single component have been moved to /components/samples
All images have been moved to /assets

andres-holguin committed:
Update UploadForm.vue

Make the Upload form a Dialog activated by a button.

andres-holguin committed:
Update Navbar.vue

Include the Update form (and its button) in the NavBar

andres-holguin committed:
Update index.html

Update homepage to use Vue components.

keelywan committed:
Change styling of search bar to work with vuetify

andres-holguin committed:
Merge branch 'aradhya' into pages

andres-holguin committed:
Create NoteGridCard.vue

First iteration of a NoteGridCard, WIP

keelywan committed:
Merge remote-tracking branch 'origin/components' into search-service

andres-holguin committed:
Update index.html

Make index the search results page.

keelywan committed:
Add test data

keelywan committed:
Make changes to components to test search service

andres-holguin committed:
Merge branch 'search-service' into pages

andres-holguin committed:
Update index.html

Make page with searchbar that loads note results.

andres-holguin committed:
Delete upload-notes-form.jsp

andres-holguin committed:
Update NotePreview.vue

Anticipate potential conflict with DB storage.

Aradhya-B committed:
Fix syntax errors in misc note service

Aradhya-B committed:
Change order of sql query execution to follow foreign key constraints

Aradhya-B committed:
Merge branch 'pages' into aradhya

andres-holguin committed:
Update Navbar.vue

Remove custom directive, already set by Vuetify

Aradhya-B committed:
Store lowercase labels in the database

Aradhya-B committed:
Make author id getString instead of getLong

keelywan committed:
Make changes to variable names on components

Aradhya-B committed:
Make author id getString instead of getLong

Aradhya-B committed:
Change static method to instance

Aradhya-B committed:
Make authorId in Note a string and note a long

Aradhya-B committed:
Change author id to string in Note constructor tests

Aradhya-B committed:
Merge branch 'search-service' into aradhya

Aradhya-B committed:
Change blobkey hardcode to aradhya local

Aradhya-B committed:
Make blob key dynamically generate from received note object

Aradhya-B committed:
Fix filepaths on navbar

andres-holguin committed:
Merge branch 'pages' into demo

andres-holguin committed:
Update Navbar.vue

User vuetify components for dropdown menu and activator

andres-holguin committed:
Update Upload Form

- Updated sample page for upload form
- Removed float: left from upload form's activator.

keelywan committed:
Change action items on navbar

keelywan committed:
Retrieve misc labels from database by note id

andres-holguin committed:
Create MockSearchButton.vue

Create a component that acts like the searchbar, emitting mock note data for UI testing.

andres-holguin committed:
Update NoteGrid.vue

Remove debug message

andres-holguin committed:
Update NoteGrid.vue

Add filter data and include as slot prop

andres-holguin committed:
Update NoteGridCollection.vue

Include functionality for filtering a NoteGridCollection's notes.

andres-holguin committed:
Update MockSearchButton and Test.html

Update for testing filter capability.

andres-holguin committed:
Update test.html

Use slot props for note-grid filter.

andres-holguin committed:
Update NoteGridCollection.vue

- Abstract the filter function to noteFilter
- Expand filter to incldude note title, school, and course

andres-holguin committed:
Update MockSearchButton.vue

Further specify mock data

andres-holguin committed:
Create FilterList.vue

Create component with filter selection for search results.

andres-holguin committed:
Update NoteGrid with FilterList

Add the FilterList component to the NoteGrid

andres-holguin committed:
Update FilterList.vue to fetch labels

The FilterList will fetch the suggested labels based on the School and Course used in the search query.

andres-holguin committed:
Add TODO to FilterList.vue

Mention that servlet does not exist yet.

andres-holguin committed:
Tweak NoteGridCollection.vue

Remove unecessary else statement.

andres-holguin committed:
Add sortFunc prop to NoteGridCollection

Add a prop to NoteGridCollection which is used to compare notes for sorting the collection.

andres-holguin committed:
Rename prop in NoteGridCollection

Rename sortFunc to compareFunc, and propogate that change.

andres-holguin committed:
Update NotePreview.vue to remove computed src

As per a PR discussion, the computed src is removed since the back end will provide a url as the pdfSrc

andres-holguin committed:
Update MockSearchButton with unique Dates

Useful for testing the sort by time feature.

andres-holguin committed:
Update MockSearchButton with new dates

Use dates relative to current time rather than Date(0)

andres-holguin committed:
Update NoteGridCollection with header

Add header to NoteGridCollection components.

andres-holguin committed:
NoteGridCollection: Abstract out filter and compare functions.

Rename the way notes are computed, simplifying by moving filters and sort comparators to methods.

andres-holguin committed:
NoteGridCollection: Add date filtering

Allow filtering by the age of a note.

andres-holguin committed:
Update test.html

Example of using new filters and sorting for NoteGridCollection

andres-holguin committed:
Update template.html to ommit unecessary tags

As per Google Style Guidelines, ommit <html>, <head>, and <body>

andres-holguin committed:
Move content test.html -> index.html

Update index.html with the test content from test.html, then delete test.html

andres-holguin committed:
Tweak index.html

Change Most Downloaded category to Most Favorited

andres-holguin committed:
Update index.html with omitted tags

Also fixed a reference to a removed function.

andres-holguin committed:
NoteGridCollection: Make filter case insensitive

andres-holguin committed:
Merge branch 'master' into search-filters-and-sorting

andres-holguin committed:
Fix some merge conflicts.

The last commit allowed some merge conflicts to go unresolved, somehow. This fixes that.

cftorres commented on src/main/webapp/components/Navbar.vue:
```diff
@@ -4,9 +4,9 @@
-             id="signin-link">
+            dark
+            @click.stop="signIn"
+            id="signin-link">
```
nit: shouldn't these be indented one more space in as before?

cftorres commented on src/main/webapp/index.html:
```diff
@@ -1,73 +1,75 @@
+                            :compare-func="descDate">
+      </note-grid-collection>
+      <note-grid-collection :note-data="searchResult"
+                            header="Hottest Notes"
```
Is the only thing distinguishing this from "Most favorited" that it's a subset of the other collection with some date restriction?  From a UX perspective, I'm not sure having both would really be necessary...?

andres-holguin replied:
Good point! The idea was to highlight notes that have been the most active recently (regardless of age), but that will probably require a lot of back end work so this is just a placeholder, really. I can remove or comment it out with an explanation.
Highlighting recent popularity is important, however, since the most relevant notes for a course will probably change as the semester progresses.

Aradhya-B commented on src/main/webapp/components/Navbar.vue:
```diff
@@ -89,8 +97,8 @@ module.exports = {
       const authRes = this.googleAuth.currentUser.get().getAuthResponse();
       const token = authRes.id_token;
       const expirationTime = authRes.expires_at;
-      fetch('/user-registration?idToken=' + token + '&exp=' + expirationTime, {
-        method: 'POST'
+      fetch('/user-signin?idToken=' + token + '&exp=' + expirationTime, {
+        method: "POST"
```
We want to keep this as a call to '/user-registration'. There is no path for '/user-signin'.

andres-holguin replied:
Ahh thanks for catching that. Definitely some files in this PR that shouldn't have been modified, hopefully I get them all with the clean up commit.

Aradhya-B reviewed (changes requested):
I'm a little confused about what exactly is going to be merged here and what conflicts will arise. Would it be possible to clean up this PR so only new things that haven't already been committed be added?

andres-holguin committed:
Undo some changes

Some changes were comitted from old branches, but which were overwritten by newer branches.

Aradhya-B commented on src/main/webapp/components/FilterList.vue:
```diff
@@ -0,0 +1,66 @@
+  },
+  methods: {
+    fetchSuggestedLabels: function() {
+      let url = new URL("/common-labels", window.location.href); // TODO: Create the servlet and rename "/common-labels"
```
nit: can remove this todo

Aradhya-B commented on src/main/webapp/components/FilterList.vue:
```diff
@@ -0,0 +1,66 @@
+        "MAC",
+        "ENC",
+        "COP",
+      ]
```
nit: now that we have a `/common-labels` path, might be a better to leave this as an empty array instead of hardcoding filters that might not exist, although it shouldn't really matter either way.

andres-holguin replied:
Coming to PR near you ;)

Aradhya-B commented on src/main/webapp/components/MockSearchButton.vue:
```diff
@@ -0,0 +1,58 @@
+        {
+          thumbnailSrc: "/assets/notes.png",
+          author: "Andres H.",
+          title: "AC set of notes",
+          date: new Date(Date.now() - 1000 * 60 * 60 * 24 * 3),
+          school: "ABC",
+          course: "COP3502",
+          miscLabels: ["A", "C"],
+          numDownloads: 123,
+          numFavorites: 456,
+          isFavorited: false,
+          pdfSource: "/assets/notes.pdf"
+        },
+        {
+          thumbnailSrc: "/assets/notes.png",
+          author: "Andres H.",
+          title: "BC set of notes",
+          date: new Date(Date.now() - 1000 * 60 * 60 * 24 * 6),
+          school: "DEF",
+          course: "MAC1147",
+          miscLabels: ["B", "C"],
+          numDownloads: 789,
+          numFavorites: 012,
+          isFavorited: false,
+          pdfSource: "/assets/notes.pdf"
+        },
+        {
+          thumbnailSrc: "/assets/notes.png",
+          author: "Andres H.",
+          title: "ABC set of notes",
+          date: new Date(Date.now() - 1000 * 60 * 60 * 24 * 10),
+          school: "GHI",
+          course: "ENC1101",
+          miscLabels: ["A", "B", "C"],
+          numDownloads: 345,
+          numFavorites: 678,
+          isFavorited: false,
+          pdfSource: "/assets/notes.pdf"
+        },
+      ]
```
Do we still need these hardcoded notes?

andres-holguin replied:
Since this component is only used for dev testing, yes.

keelywan commented on src/main/webapp/components/NoteGridCard.vue:
```diff
@@ -36,6 +36,12 @@
+        if(this.dateCreated) {
+          return this.dateCreated.toDateString();
+        }
+        return "";
```
Is `dateString` necessary because I think `dateCreated` displays the formatted date already?

keelywan replied:
Also, I think I forgot to add it but same thing in NotePreview.vue.

andres-holguin replied:
I'm going to make a change to dateCreated to ensure it's a parsed Date() object, since
- A date object is what the servlet responds with
- It's better for sorting/filtering notes by date.

The formatting should be done at the last possible moment.

keelywan reviewed (approved):
Looks good!

Aradhya-B reviewed (approved):
LGTM :) Great job!
